src/Utilities/Compiler/WellKnownTypeProvider.cs

```diff
@@ -126,5 +191,71 @@ internal bool IsTaskOfType([NotNullWhen(returnValue: true)] ITypeSymbol? typeSym
                 && namedTypeSymbol.TypeArguments.Length == 1
                 && typeArgumentPredicate(namedTypeSymbol.TypeArguments[0]);
         }
+
+        private static string[] GetNamespaceNamesFromFullTypeName(string fullTypeName)
+        {
+            int dotCount = 0;
+            int lastDotIndex = -1;
+            for (int i = 0; i < fullTypeName.Length; i++)
+            {
+                switch (fullTypeName[i])
+                {
+                    case '.':
+                        lastDotIndex = i;
+                        dotCount++;
+                        break;
+
+                    // Dunno if Compilation.GetTypeByMetadataName() or IAssemblySymbol.GetTypeByMetadataName() will ever 
+                    // support something like "System.Collections.Generic.List<System.String>", but if they do, we'll be ready.
+                    case '`':
+                    case '+':
+                    case '<':
+                    case '>':
+                    case '[':
+                    case ']':
+                    case '*':
+                    case ',':
+                        goto ExitLoop;
+                }
+            }
+
+            ExitLoop:
+            if (dotCount == 0)
+            {
+                return Array.Empty<string>();
+            }
+
+            string[] namespaceNames = new string[dotCount];
+            int namespaceIndex = 0;
+            int prevStartIndex = 0;
+            for (int i = 0; i <= lastDotIndex; i++)
```

Comment:
We can potentially avoid the second loop if we use and update an array builder for namespaceNames and a string builder for current namespace name in the prior loop, promoting the current namespace name into `namespaceNames` whenever a `.` is encountered.